javascripts/2048.js

```diff
@@ -1,32 +1,406 @@
+
+function addTile() {
+  var tile = $("<div data-row='' data-col='' data-val=''></div>");
+  var dataVal = Math.random() < 0.04 ? 4 : 2;
+  var emptySpaces = collectEmptySpaces(); // guard for if no empty spaces?
+
+  // function to pick row and column number
+  var randomLocation = emptySpaces[Math.floor(Math.random() * emptySpaces.length)];
+
+  tile.addClass("tile");
+  tile.attr("data-row", randomLocation[0]);
+  tile.attr("data-col", randomLocation[1]);
+  tile.attr("data-val", dataVal);
+  tile.text(dataVal);
+  tile.addClass("new");
+
+  $("#gameboard").append(tile);
+  // remove new class after animation is complete
+  tile.on("animationend", function() {
+    tile.removeClass("new");
+  })
+}
+
+function makeTurn(direction) {
+  var tiles = $(".tile"); // in order by the html
+  var turnType = rowOrColumn(direction);
+  var turnOppositeType = (turnType == "data-row") ? "data-col" : "data-row";
+  // set the positive or negative magnitude according to direction
+  var turnMagnitude = parseInt(getMagnitude(direction));
+  var score = parseInt($("#score").attr("data-score"));
+
+  // take snapshot of board at start of turn
+  var startingGameboard = gameboardSnapshot($(".tile"));
+
+  function gameboardSnapshot(tiles) {
+    var gameboard = [];
+    for (var i = 0; i < tiles.length; i++) {
+      var dataRow = tiles[i].getAttribute("data-row");
+      var dataCol = tiles[i].getAttribute("data-col");
+      var dataVal = tiles[i].getAttribute("data-val");
+      var tileProperties = new Array(dataRow, dataCol, dataVal);
+      gameboard[i] = tileProperties;
+    }
+    return gameboard;
+  }
+
+  function validMove() {
+    var validity = true;
+    var currentGameboard = gameboardSnapshot($(".tile"));
+
+    if (startingGameboard.equals(currentGameboard)) {
+      validity = false;
+    }
+    console.log("validity: " + validity);
+    return validity;
+  }
+
+  var updateScore = function(points) {
+    score += points;
+    var scoreBoard = $("#score");
+    scoreBoard.attr("data-score", score);
+    scoreBoard.text(score);
+  }
+
+  function moveTiles() {
+    var moveInDirection = makeMovement(direction);
+
+    // order tiles correctly
+    var sortedTiles = orderTiles();
+
+    for (var i = 0; i < sortedTiles.length; i++) {
+      moveInDirection(sortedTiles[i]);
+    }
+  }
+
+  function orderTiles() {
+    // check if movement is up/down (row) or right/left (col)
+    // check if movement makes values get smaller (up/left) or bigger (down/right)
+    var magnitude = parseInt(getMagnitude(direction));
+    // store type of non-movement dimension
+
+    // split into groups by type
+    var typeA = [];
+    var typeB = [];
+    var typeC = [];
+    var typeD = [];
+
+    var upLeftOrder = [1, 2, 3, 4];
+    var downRightOrder = [4, 3, 2, 1];
+
+    var order = (direction == 38 || direction == 37) ? upLeftOrder : downRightOrder;
+
+    for (var i = 0; i < tiles.length; i++) {
+      if (tiles[i].getAttribute(turnType) == order[0]) {
```

Comment:
Suggestion: do not call `tiles[i].getAttribute(turnType)` 4 times. Store it in a variable instead for readability and reduce processing time (if `tiles` had thousands of items it could become slow).